cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/linkanalysis/Hits.scala

```diff
@@ -0,0 +1,74 @@
+package com.twitter.cassovary.algorithms.linkanalysis
+
+import com.twitter.cassovary.graph.{Node, DirectedGraph}
+
+/**
+ * Stores all parameters for Hits algorithm
+ * @param maxIterations The maximum number of times that the link analysis algorithm will run before termination
+ * @param tolerance The maximum error allowed.
+ * @param normalize Flag `true` to return normalized values
+ */
+case class HitsParams(maxIterations: Option[Int] = Some(100),
+                      tolerance: Double = 1.0e-8,
+                      normalize: Boolean = true)
+  extends Params
+
+/**
+ * Stores all values necessary to fully describe one Hits iteration
+ * @param hubs Array of values indexed by node id storing hubs values for each node
+ * @param authorities Array of values indexed by node id storing authorities values for each node
+ * @param error  The T1 error for the current iteration vs the previous iteration.
+ */
+case class HitsIterationState(hubs: Array[Double],
+                         authorities: Array[Double],
+                         iteration: Int,
+                         error: Double)
+  extends IterationState
+
+/**
+ * Case class for performing Hits algorithm.  Hits is a link analysis algoritm that returns two values characterizing
+ * each node.  Each node receives both a hub value and an authority value.  A node that is characterized by a large
+ * hub value is one that has many high quality outbound connections to other nodes, while a node that is characterized
+ * by a large authority value has inbound connections from high quality hub nodes.
+ * @param params The set of all parameters passed into our algorithm
+ */
+class Hits(graph: DirectedGraph[Node], params: HitsParams)
+  extends AbstractLinkAnalysis[HitsIterationState](graph, params, "hits") {
+  
+  private val normalize = params.normalize
+
+  private def scale(m: Array[Double], byMax: Boolean): Array[Double] = {
+    val d = if (byMax) m.max else m.sum
+    if (d <= 0) m else m.map { v => v / d }
+  }
+
+  protected def defaultInitialIteration: HitsIterationState = {
+    val hubs = new Array[Double](graph.maxNodeId + 1)
+    val authorities = new Array[Double](graph.maxNodeId + 1)
+    
+    graph foreach { n => hubs(n.id) = 1.0 / graph.nodeCount }
+    new HitsIterationState(hubs, authorities, 0, 100 + tolerance)
+  }
+  
+  def iterate(prevIteration: HitsIterationState): HitsIterationState = {
+    val beforeHubs = prevIteration.hubs
+
+    val partialAuth, partialHubs = new Array[Double](graph.maxNodeId + 1)
+    graph foreach { node =>
```

Comment:
@pankajgupta In working on this, I noticed something that didn't quite make sense to me--before submitting an issue, I wanted to make sure this was an actual bug.  

In TestGraphs, `g6` is created using 
```scala  
val nodeSeqIterator = Seq(
      NodeIdEdgesMaxId(10, Array(11, 12, 13)),
      NodeIdEdgesMaxId(11, Array(12, 14)),
      NodeIdEdgesMaxId(12, Array(14)),
      NodeIdEdgesMaxId(13, Array(12, 14)),
      NodeIdEdgesMaxId(14, Array(15)),
      NodeIdEdgesMaxId(15, Array(10, 11))
      )
```
which, according to the documentation, should be an iterator of `NodeIdEdgesMaxId` objects with `nodeId` as the first constructor arg and an array of out edges as the second arg (seen [here](https://github.com/twitter/cassovary/blob/master/cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDirectedGraph.scala#L36-L39).

So when `g6_onlyin` and `g6_onlyout` are created [here](https://github.com/twitter/cassovary/blob/master/cassovary-core/src/main/scala/com/twitter/cassovary/graph/TestGraph.scala#L99-L102), they _should_ represent the same graph but only differ in node storage strategy.

In the console, if you print out `g6`, you get the following:
```scala
both: com.twitter.cassovary.graph.ArrayBasedDirectedGraph =
Node count: 6
Edge count: 11
Nodes:
NodeId => 10
InboundNodes[1] =>15|
OutboundNodes[3] =>11|12|13|

NodeId => 11
InboundNodes[2] =>10|15|
OutboundNodes[2] =>12|14|

NodeId => 12
InboundNodes[3] =>10|11|13|
OutboundNodes[1] =>14|

NodeId => 13
InboundNodes[1] =>10|
OutboundNodes[2] =>12|14|

NodeId => 14
InboundNodes[3] =>11|12|13|
OutboundNodes[1] =>15|

NodeId => 15
InboundNodes[1] =>14|
OutboundNodes[2] =>10|11|
```
`g6_onlyout` prints:
```scala
out: com.twitter.cassovary.graph.ArrayBasedDirectedGraph =
Node count: 6
Edge count: 11
Nodes:
NodeId => 10
InboundNodes[0] =>
OutboundNodes[3] =>11|12|13|

NodeId => 11
InboundNodes[0] =>
OutboundNodes[2] =>12|14|

NodeId => 12
InboundNodes[0] =>
OutboundNodes[1] =>14|

NodeId => 13
InboundNodes[0] =>
OutboundNodes[2] =>12|14|

NodeId => 14
InboundNodes[0] =>
OutboundNodes[1] =>15|

NodeId => 15
InboundNodes[0] =>
OutboundNodes[2] =>10|11|
```
which is expected.  *BUT* `g6_onlyin` has a very strange behavior:
```scala
in: com.twitter.cassovary.graph.ArrayBasedDirectedGraph =
Node count: 6
Edge count: 11
Nodes:
NodeId => 10
InboundNodes[3] =>11|12|13|
OutboundNodes[0] =>

NodeId => 11
InboundNodes[2] =>12|14|
OutboundNodes[0] =>

NodeId => 12
InboundNodes[1] =>14|
OutboundNodes[0] =>

NodeId => 13
InboundNodes[2] =>12|14|
OutboundNodes[0] =>

NodeId => 14
InboundNodes[1] =>15|
OutboundNodes[0] =>

NodeId => 15
InboundNodes[2] =>10|11|
OutboundNodes[0] =>
```
It seems as though `g6_onlyin` has just taken the outbound edges and stored them as inbound edges which means that `g6_onlyin` and `g6_onlyout` do not represent the same graph.  

Just wanted a sanity check before I submit an issue.